src/main/java/org/roaringbitmap/longlong/Roaring64NavigableMap.java

```diff
@@ -0,0 +1,1351 @@
+
+      final long previousBucketCardinality;
+      if (highPosition == 0) {
+        previousBucketCardinality = 0;
+      } else {
+        previousBucketCardinality = sortedCumulatedCardinality[highPosition - 1];
+      }
+
+      BitmapDataProvider lowBitmap = highToBitmap.get(sortedHighs[highPosition]);
+
+      // Rank is previous cardinality plus rank in current bitmap
+      return previousBucketCardinality + lowBitmap.rankLong(low);
+    } else {
+      // There is no bucket holding this item: insertionPoint is previous bitmap
+      int insertionPoint = -highPosition - 1;
+
+      if (insertionPoint == 0) {
+        // this key is before all inserted keys
+        return 0;
+      } else {
+        // The rank is the cardinality of this previous bitmap
+        return sortedCumulatedCardinality[insertionPoint - 1];
+      }
+    }
+  }
+
+  // https://github.com/RoaringBitmap/CRoaring/blob/master/cpp/roaring64map.hh
+  private long rankLongNoCache(int high, int low) {
+    long result = 0L;
+
+    BitmapDataProvider lastBitmap = highToBitmap.get(high);
+    if (lastBitmap == null) {
+      // There is no value with same high: the rank is a sum of cardinalities
+      for (Entry<Integer, BitmapDataProvider> bitmap : highToBitmap.entrySet()) {
+        if (bitmap.getKey().intValue() > high) {
+          break;
+        } else {
+          result += bitmap.getValue().getLongCardinality();
+        }
+      }
+    } else {
+      for (BitmapDataProvider bitmap : highToBitmap.values()) {
+        if (bitmap == lastBitmap) {
+          result += bitmap.rankLong(low);
+          break;
+        } else {
+          result += bitmap.getLongCardinality();
+        }
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * 
+   * @param high for which high bucket should we compute the cardinality
+   * @return the highest validatedIndex
+   */
+  protected int ensureCumulatives(int high) {
+    if (allValid) {
+      // the whole array is valid (up-to its actual length, not its capacity)
+      return highToBitmap.size();
+    } else if (compare(high, firstHighNotValid) < 0) {
+      // The high is strictly below the first not valid: it is valid
+
+      // sortedHighs may have only a subset of valid values on the right. However, these invalid
+      // values have been set to maxValue, and we are here as high < firstHighNotValid ==> high <
+      // maxHigh()
+      int position = binarySearch(sortedHighs, high);
+
+      if (position >= 0) {
+        // This high has a bitmap: +1 as this index will be used as right (excluded) bound in a
+        // binary-search
+        return position + 1;
+      } else {
+        // This high has no bitmap: it could be between 2 highs with bitmaps
+        int insertionPosition = -position - 1;
+        return insertionPosition;
+      }
+    } else {
+      // For each deprecated buckets
+      SortedMap<Integer, BitmapDataProvider> tailMap =
+          highToBitmap.tailMap(firstHighNotValid, true);
```

Comment:
Can you consider combining `#tailMap(...)` to `#headMap(...)` so you only consider the relevant portion of the map ?
This requires that you have another way to compute `indexOf`. 
But the advantage is that you can iterate in reverse order and predict once for all how much you may need to grow `sortedHighs`

Reply:
I feel the relevant portion of actually the tail: we need to iterate along the tail to compute the missing cardinalities. Then I guess I miss your point here.

Anyway, `sortedHighs` resizing could take in account `highToBitmap` size (which is free for a `TreeMap`), and not depends on `firstHighNotValid` at all. Thanks for pointing this out.

I have difficulties to find a way to compute indexOf differently. I added a note about the feasibility to compute it from sortedHighs. `tailMap.size()` should not be called anyway.

Reply:
Good idea 😄 . You can try as well `#lastEntry`, which will give you the very last value to insert, at the cost of going through the highest branches.
But this was just a suggestion. I guess you would have spent time on this if it appeared to be a bottleneck in your benchmarks.